# Imports

In [57]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [58]:
pd.set_option('display.float_format', lambda x: '%.2f' % x) 

In [59]:
videos_stats = pd.read_csv('videos_stats.csv')
videos_stats

Title     Video ID  \
0     Apple Pay Is Killing the Physical Wallet After...  wAZZ-UWGVHI   
1                       The most EXPENSIVE thing I own.  b3x28s61q3c   
2                    My New House Gaming Setup is SICK!  4mgePWWCAmA   
3     Petrol Vs Liquid Nitrogen | Freezing Experimen...  kXiYSI7H2b0   
4                        Best Back to School Tech 2022!  ErMwWXQxHp0   
...                                                 ...          ...   
1876                 Should You Learn Machine Learning?  AO6urf07KjE   
1877            Todos podemos aprender Machine learning  7ClLKBUvmRk   
1878  Andrew Ng: Deep Learning, Education, and Real-...  0jspaMLxBig   
1879                          What is Machine Learning?  f_uwKZIAeM0   
1880                   Don&#39;t learn machine learning  cyLWtMSry58   

     Published At           Keyword    Likes  Comments      Views  
0      23/08/2022              tech  3407.00    672.00  135612.00  
1      24/08/2022              tech 76779.00   4306.00 1758063.00  
2      23/08/2022              tech 63825.00   3338.00 1564007.00  
3      23/08/2022              tech 71566.00   1426.00  922918.00  
4      08/08/2022              tech 96513.00   5155.00 1855644.00  
...           ...               ...      ...       ...        ...  
1876   14/06/2021  machine learning 10259.00    416.00  386360.00  
1877   08/10/2017  machine learning  2981.00     72.00  431421.00  
1878   20/02/2020  machine learning  5198.00    443.00  226152.00  
1879   11/01/2017  machine learning      NaN       NaN        NaN  
1880   30/03/2020  machine learning      NaN       NaN        NaN  

[1881 rows x 7 columns]

In [60]:
comments = pd.read_csv('comments.csv')
comments

Video ID                                            Comment  Likes  \
0      wAZZ-UWGVHI  Let's not forget that Apple Pay in 2014 requir...     95   
1      wAZZ-UWGVHI  Here in NZ 50% of retailers don’t even have co...     19   
2      wAZZ-UWGVHI  I will forever acknowledge this channel with t...    161   
3      wAZZ-UWGVHI  Whenever I go to a place that doesn’t take App...      8   
4      wAZZ-UWGVHI  Apple Pay is so convenient, secure, and easy t...     34   
...            ...                                                ...    ...   
18404  cyLWtMSry58  I really like the point about engineering tool...      0   
18405  cyLWtMSry58  I’ve just started exploring this field. And th...     20   
18406  cyLWtMSry58  Excelente video con una pregunta filosófica pr...      1   
18407  cyLWtMSry58  Hey Daniel, just discovered your channel a cou...     35   
18408  cyLWtMSry58  This is great. Focus is key. A playful approac...      0   

       Sentiment  
0              1  
1              0  
2              2  
3              0  
4              2  
...          ...  
18404          2  
18405          2  
18406          1  
18407          2  
18408          2  

[18409 rows x 4 columns]

# Cleaning the data

looking for cells with invalid values

In [61]:
videos_stats.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1881 entries, 0 to 1880
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Title         1881 non-null   object 
 1   Video ID      1881 non-null   object 
 2   Published At  1881 non-null   object 
 3   Keyword       1881 non-null   object 
 4   Likes         1879 non-null   float64
 5   Comments      1879 non-null   float64
 6   Views         1879 non-null   float64
dtypes: float64(3), object(4)
memory usage: 103.0+ KB


In [62]:
videos_stats.isna().sum()

Title           0
Video ID        0
Published At    0
Keyword         0
Likes           2
Comments        2
Views           2
dtype: int64

In [63]:
videos_stats = videos_stats.dropna()

Checking if there is invalid values for the likes and the comments

In [64]:
videos_stats[(videos_stats['Likes']<=0) | ( videos_stats['Comments']<=0) ] 

Title     Video ID  \
243   How To Build A Business That Works | Brian Tra...  MN7yfV4UuCI   
266   ตลาดถุงมือยางทรุด! ฉุดราคาร่วงลงทุนหด | BUSINE...  cG37cEi1nPc   
583                         DON&#39;T GOOGLE THIS STUFF  NAV2laKrDv4   
1016  15 Kilo Zayıf Gösteren Kız Makyajı |Arkadaşlar...       #NAME?   
1393  Mathematics and Chemistry : MathChemistry.com ...  V_vguZj_7FE   
1677  FULL Marvel Studios Panel from Hall H | San Di...  rf-NqJQJHBU   
1701  SURVIVOR 🦀🦂 Best Action Movies 2022 🦀🦂 Latest ...  __7MkaWFObQ   
1709  BEST Auditions Of Songs From Movies | Amazing ...  ZcBBGC8_mfU   
1710  The Magnificent Seven - Full Movie In English ...  1MZY_0tYmrs   
1711  TOP SECRET MISSION ⭐🌟 Best Action Movies 2022 ...  PIFa7wJ9c0c   
1714  Vikram Vedha Movie Teaser Review | KRK | #krkr...  2FYvHn12pOQ   

     Published At        Keyword    Likes  Comments      Views  
243    09/02/2021       business    -1.00   1144.00  676300.00  
266    23/08/2022       business    -1.00     13.00   12332.00  
583    28/08/2017         google    -1.00  18676.00 6515548.00  
1016   23/08/2022        mukbang    -1.00    335.00   27451.00  
1393   15/04/2013  mathchemistry     0.00      1.00      25.00  
1677   24/07/2022         marvel    -1.00    821.00  553892.00  
1701   20/08/2022         movies   105.00     -1.00   28219.00  
1709   25/07/2022         movies 20138.00     -1.00 5351960.00  
1710   29/12/2021         movies    -1.00    180.00 2389222.00  
1711   18/07/2022         movies   643.00      0.00   24012.00  
1714   24/08/2022         movies    -1.00    105.00 1541874.00

There is 11 entries with wrong values in either likes and comments 
we can fix it by first checking if those videos are actuall existing videos by searching or validating using the video Id which is the link of the video but, they are few so it's unecceary, we will drop them

In [65]:
videos_stats = videos_stats[(videos_stats['Likes']>0) & ( videos_stats['Comments']>0) ]

Checking the datatypes

In [93]:
videos_stats.dtypes

Title                   object
Video ID                object
Published At    datetime64[ns]
Keyword                 object
Likes                    int64
Comments                 int64
Views                    int64
dtype: object

In [67]:
videos_stats['Published At'] = pd.to_datetime(videos_stats ['Published At'], format='%d/%m/%Y', errors='coerce')

changing the data type of  views, likes and comments to int

In [91]:
# transforming columns from float to int64
videos_stats['Views'] = videos_stats['Views'].astype('int64')
videos_stats['Likes'] = videos_stats['Likes'].astype('int64')
videos_stats['Comments'] = videos_stats['Comments'].astype('int64')

In [92]:
videos_stats.describe()

Likes  Comments         Views
count     1868.00   1868.00       1868.00
mean    171051.27   7898.25   11672135.57
std     798466.37  37986.47  108761106.89
min          1.00      1.00         63.00
25%       2705.25    201.00      85156.25
50%      15011.00    820.00     591947.00
75%      61091.75   3411.50    2814055.25
max   16445558.00 732818.00 4034122271.00

In [94]:
# Check for empty or whitespace-only titles
empty_titles = videos_stats['Title'].str.strip() == ''

# Check for non-string titles
non_string_titles = videos_stats['Title'].apply(lambda x: not isinstance(x, str))


In [95]:
videos_stats[empty_titles ]

Empty DataFrame
Columns: [Title, Video ID, Published At, Keyword, Likes, Comments, Views]
Index: []

In [96]:
videos_stats[non_string_titles]

Empty DataFrame
Columns: [Title, Video ID, Published At, Keyword, Likes, Comments, Views]
Index: []

Also we will find in some titles text like "& #39;" which  HTML entity representing the apostrophe character

The ID length should be 11 characters,so if we checked for enteries that doesn't met this critaria, we will find some with missing video Id, but they are still valid videos so we will keep it

In [97]:
invalid_length_video_ids = videos_stats['Video ID'].apply(lambda x: len(x) != 11)
videos_stats[invalid_length_video_ids]

Title Video ID Published At  \
36                              My Everyday Tech: 2022!   #NAME?   2022-01-07   
44    Sushi Chef Answers Sushi Questions From Twitte...   #NAME?   2022-05-26   
45    Futurum Research&#39;s Daniel Newman breaks do...   #NAME?   2022-08-24   
48    How Biden&#39;s student loan forgiveness progr...   #NAME?   2022-08-24   
71             Deadly car bomb detonates outside Moscow   #NAME?   2022-08-22   
105        🎧 FOLLOW ME ON LOCO | ACHANAK BAYANAK GAMING   #NAME?   2022-08-24   
119                   New PS4 &amp; PS5 Games This Week   #NAME?   2022-08-23   
157             All Sports Golf Battle 2 | Dude Perfect   #NAME?   2017-12-04   
238   DON&#39;T USE DOMAIN.COM Before Watch THIS VID...   #NAME?   2022-08-01   
493   Champions Chess Tour: FTX Crypto Cup | Day 2 |...   #NAME?   2022-08-16   
509   How to Assess ANY Chess Position: The Ultimate...   #NAME?   2022-08-24   
588   ZOMBIE GIRL ESCAPE PREGNANCY PRANK BATTLE - Do...   #NAME?   2020-11-28   
614   English Job Interview Dos &amp; Dont&#39;s! | ...   #NAME?   2021-08-03   
664   35 Rare Shocking and Heartbreaking Historical ...   #NAME?   2022-08-23   
667   The History Of Earth&#39;s Five Mass Extinctio...   #NAME?   2022-04-13   
853   ASMR for Those Who Want to Sleep Soundly Now /...   #NAME?   2021-12-02   
855   ASMR Relaxing Spa Facial Treatment 💆🏽‍♀️ (Pers...   #NAME?   2022-08-09   
898   if someone asks you to play music, play this p...   #NAME?   2022-02-08   
986   ASMR MUKBANG DOUBLE BIG MAC &amp; CHEESY HASH ...   #NAME?   2020-04-18   
1062  NO MORE A FAMILY FRIENDLY VLOGGER | LAKSHAY CH...   #NAME?   2022-08-23   
1095            alone with my thoughts. 🧺 calm lofi mix   #NAME?   2022-06-07   
1146  ALL MY EXOTIC ANIMALS IN ONE VIDEO ! FULL ZOO ...   #NAME?   2022-07-30   
1195  I found SECRET ROAD to OBUNGA PLANET in MINECR...   #NAME?   2022-08-23   
1492           100+ Computer Science Concepts Explained   #NAME?   2022-05-04   
1525  CS50 FULL REVIEW  - Best Course to Learn Compu...   #NAME?   2022-02-08   
1583              Nintendo Switch accessories you NEED!   #NAME?   2022-08-22   
1686  Lifetime Movies (2022) #LMN Movies ~ New Lifet...   #NAME?   2022-08-20   
1733  Data Science Full Course - Learn Data Science ...   #NAME?   2019-08-18   
1791  Making Your First Indie Game (5 Tips for 2021 ...   #NAME?   2021-11-10   

               Keyword   Likes  Comments      Views  
36                tech  134194      6102    4098623  
44                tech   77411      2098    2474364  
45                tech      24        45       3461  
48                news    1029      2347      97434  
71                news    6379      4853     808787  
105             gaming    1525        36      14877  
119             gaming     789        77      19223  
157             sports  867074     43024  106014469  
238           business     114         2       2270  
493              chess    1234        44      92760  
509              chess      58        21        601  
588             google  349923      4082  122994457  
614          interview   19630       515     621093  
664            history     160        10       8135  
667            history   11595      2294    1487029  
853               asmr  454036      6622   23568994  
855               asmr   34526      1668    1018753  
898              music   94131       920    3921128  
986            mukbang  378858     18860   17975269  
1062          reaction  196898     15525    1223879  
1095              lofi    9965       197     325200  
1146           animals    6473       455     267705  
1195         minecraft    1333        22     119487  
1492  computer science   55719      1606     802372  
1525  computer science   10409       385     199395  
1583          nintendo   24496       257     254500  
1686            movies    1810        13     285609  
1733      data science  118846      3501    3069097  
1791  game development   18486       733     353784

In [98]:
# saving the video stats to  a new csv file called cleaned_video_stats.csv
videos_stats.to_csv('cleaned_video_stats.csv', index=False)


# Exploratory Data Analysis (EDA)

## Summary statistics

In [99]:
videos_stats.describe()

Likes  Comments         Views
count     1868.00   1868.00       1868.00
mean    171051.27   7898.25   11672135.57
std     798466.37  37986.47  108761106.89
min          1.00      1.00         63.00
25%       2705.25    201.00      85156.25
50%      15011.00    820.00     591947.00
75%      61091.75   3411.50    2814055.25
max   16445558.00 732818.00 4034122271.00

The dataset provides the following insights from the summary statistics:

The dataset contains 1,868 videos.

Likes: The number of likes ranges from 1 to over 16 million, with a mean of approximately 171,000 likes, with a high standard deviation indicating a wide range of view counts.

Comments: The number of comments ranges from 1 to over 732,000, with a mean of approximately 7,898 comments.

Views: The number of views ranges from 63 to over 4 billion, with a mean of approximately 11.67 million views.

The median values for views, likes, and comments are significantly lower than the mean, suggesting that a few videos with extremely high engagement skew the data.

In [120]:
# Group by keyword and calculate mean engagement metrics
keyword_engagement = videos_stats.groupby('Keyword').mean()[['Views', 'Likes', 'Comments']].reset_index()

keyword_engagement.sort_values(by='Views', ascending=False)


C:\Users\USCS\AppData\Local\Temp\ipykernel_9072\2800320317.py:2: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  keyword_engagement = videos_stats.groupby('Keyword').mean()[['Views', 'Likes', 'Comments']].reset_index()


Keyword        Views      Likes  Comments
17            google 105566093.91  481875.86  22736.43
0            animals  94723960.92  760775.79  21871.82
28           mrbeast  66764003.98 2105914.36  95944.48
3                bed  53893228.61  473682.18  15180.95
30             music  29364893.26  314188.11  12288.93
18           history  15047130.12  273410.66  20190.02
10             cubes  15038739.34  303061.20   6833.08
29           mukbang  11151984.20  144620.45   7613.11
1              apple  10746930.45  118078.90   8618.86
36            sports   8601204.73   90518.88   3828.02
19            how-to   7809284.92  158898.15   6951.44
5           business   7536008.37   75114.93   1895.26
39          tutorial   6761032.02  167720.88   7935.40
24            marvel   6737756.86  214672.67   9881.90
14              food   5252406.25   95014.65   5523.21
27            movies   5153833.78   34097.80   3886.47
4            biology   4121605.83   66483.79   4292.45
22              lofi   3989972.22   73436.00   2399.93
33           physics   3692387.28   98700.90   4441.20
25     mathchemistry   3565846.71  100552.14   5752.21
20         interview   2966111.86   73326.94   5428.24
12         education   2684432.83   62292.33   4113.46
37              tech   1918806.94   60534.42   3152.17
26         minecraft   1812770.74   61179.56   2843.72
2               asmr   1706615.82   36615.30   1227.30
40              xbox   1703980.92   38820.06   2870.43
32          nintendo   1665042.25   44573.54   3694.73
38          trolling   1420141.02   61387.04   3056.82
6              chess   1323293.96   36042.38   1472.32
23  machine learning   1312602.25   35278.06   1235.44
8   computer science   1191958.71   33663.54   1170.77
35               sat   1065868.71   31030.80   2027.49
21        literature    863021.24   17825.41    879.85
15  game development    724688.54   26178.68   1376.34
13           finance    694223.44   13235.49   1084.03
34          reaction    623172.46   73959.78   3246.72
7                cnn    554240.38    8835.74   7574.74
11      data science    544771.98   17175.78    517.52
16            gaming    519645.49   24897.28   1376.91
9             crypto    404608.22    8521.36    546.62
31              news    247492.18    3013.10   1183.44